src/Diff/Internal/FingerprintPatcher.php

```diff
@@ -41,30 +37,164 @@ public function patchFingerprint( Fingerprint $fingerprint, EntityDiff $patch )
 		}
 	}
 
+	/**
+	 * @param AliasGroupList $groups
+	 * @param Diff $patch
+	 *
+	 * @throws PatcherException
+	 */
 	private function patchAliasGroupList( AliasGroupList $groups, Diff $patch ) {
-		$original = $groups->toTextArray();
-		$patched = $this->patcher->patch( $original, $patch );
+		foreach ( $patch as $lang => $diffOp ) {
+			$this->patchAliasGroup( $groups, $lang, $diffOp );
+		}
+	}
+
+	/**
+	 * @see MapPatcher
+	 *
+	 * @param AliasGroupList $groups
+	 * @param string $lang
+	 * @param DiffOp $diffOp
+	 *
+	 * @throws PatcherException
+	 */
+	private function patchAliasGroup( AliasGroupList $groups, $lang, DiffOp $diffOp ) {
+		$hasLang = $groups->hasGroupForLanguage( $lang );
 
-		foreach ( $patched as $languageCode => $aliases ) {
-			$groups->setAliasesForLanguage( $languageCode, $aliases );
+		if ( $diffOp instanceof DiffOpAdd ) {
+			if ( !$hasLang ) {
+				$groups->setAliasesForLanguage( $lang, $diffOp->getNewValue() );
+			}
+		} elseif ( $diffOp instanceof DiffOpChange ) {
+			$this->applyAliasGroupChange( $groups, $lang, $diffOp );
+		} elseif ( $diffOp instanceof DiffOpRemove ) {
+			if ( $hasLang
+				&& $groups->getByLanguage( $lang )->getAliases() === $diffOp->getOldValue()
+			) {
+				$groups->removeByLanguage( $lang );
+			}
+		} elseif ( $diffOp instanceof Diff ) {
+			$this->applyAliasGroupDiff( $groups, $lang, $diffOp );
+		} else {
+			throw new PatcherException( 'Invalid aliases diff' );
 		}
+	}
+
+	/**
+	 * @param AliasGroupList $groups
+	 * @param string $lang
+	 * @param DiffOpChange $patch
+	 */
+	private function applyAliasGroupChange( AliasGroupList $groups, $lang, DiffOpChange $patch ) {
+		if ( $groups->hasGroupForLanguage( $lang )
+			&& $groups->getByLanguage( $lang )->getAliases() === $patch->getOldValue()
+		) {
+			$groups->setAliasesForLanguage( $lang, $patch->getNewValue() );
+		}
+	}
+
+	/**
+	 * @param AliasGroupList $groups
+	 * @param string $lang
+	 * @param Diff $patch
+	 */
+	private function applyAliasGroupDiff( AliasGroupList $groups, $lang, Diff $patch ) {
+		$hasLang = $groups->hasGroupForLanguage( $lang );
+
+		if ( $hasLang || !$this->containsOperationsOnOldValues( $patch ) ) {
+			$aliases = $hasLang ? $groups->getByLanguage( $lang )->getAliases() : array();
+			$aliases = $this->getPatchedAliases( $aliases, $patch );
+			$groups->setAliasesForLanguage( $lang, $aliases );
+		}
+	}
 
-		foreach ( array_diff_key( $original, $patched ) as $languageCode => $aliases ) {
-			$groups->removeByLanguage( $languageCode );
+	/**
+	 * @param Diff $diff
+	 *
+	 * @return bool
+	 */
+	private function containsOperationsOnOldValues( Diff $diff ) {
+		return $diff->getChanges() !== array()
+			|| $diff->getRemovals() !== array();
+	}
+
+	/**
+	 * @see ListPatcher
+	 *
+	 * @param string[] $aliases
+	 * @param Diff $patch
+	 *
+	 * @throws PatcherException
+	 * @return string[]
+	 */
+	private function getPatchedAliases( array $aliases, Diff $patch ) {
```

Comment:
Note: This whole method could be replaced with a `ListPatcher` call, but this needs https://github.com/wmde/Diff/pull/65.